stl/inc/xutility

```diff
@@ -4650,55 +4650,53 @@ _FwdIt fill_n(_ExPo&&, _FwdIt _Dest, _Diff _Count_raw, const _Ty& _Val) noexcept
-          _Value_equality_is_bitwise_equality<_Elem1, _Elem2>>> {
-    // allow memcmping same-size integral non-bool non-volatile bitwise types using equal_to<>
-};
+_INLINE_VAR constexpr bool _Equal_memcmp_is_safe_helper<_Elem1, _Elem2,
+    enable_if_t<sizeof(_Elem1) == sizeof(_Elem2) //
+                    && _Is_nonbool_integral<_Elem1> && _Is_nonbool_integral<_Elem2> //
+                    && !is_volatile_v<_Elem1> && !is_volatile_v<_Elem2>,
+        equal_to<>>> = _Value_equality_is_bitwise_equality<_Elem1, _Elem2>;
 
+// allow memcmping pointers-to-cv-T with equal_to<>
 template <class _Elem1, class _Elem2>
-struct _Equal_memcmp_is_safe_helper<_Elem1*, _Elem2*, equal_to<>>
-    : is_same<remove_cv_t<_Elem1>, remove_cv_t<_Elem2>>::type {}; // allow memcmping pointers-to-cv-T with equal_to<>
+_INLINE_VAR constexpr bool _Equal_memcmp_is_safe_helper<_Elem1*, _Elem2*, equal_to<>> =
+    is_same_v<remove_cv_t<_Elem1>, remove_cv_t<_Elem2>>;
 
 #ifdef __cpp_lib_byte
+// allow memcmping std::byte
+// inline is required here as explicit specialization of variable templates are problematic in C++14 and below.
+// However, std::byte is C++17 and above so we are safe
 template <>
-struct _Equal_memcmp_is_safe_helper<byte, byte, equal_to<>> : true_type {}; // allow memcmping std::byte
+inline constexpr bool _Equal_memcmp_is_safe_helper<byte, byte, equal_to<>> = true;
 #endif // __cpp_lib_byte
 
+// Treat equal_to with exact T as equal_to<>.
+// This is safe because we only activate the optimization for builtin _Elem (and std::byte).
 template <class _Elem>
-struct _Equal_memcmp_is_safe_helper<_Elem, _Elem, equal_to<_Elem>>
-    : _Equal_memcmp_is_safe_helper<_Elem, _Elem, equal_to<>>::type {
-    // treat equal_to with exact T as equal_to<> this is safe because we only activate the optimization for builtin
-    // _Elem (and std::byte)
-};
+_INLINE_VAR constexpr bool _Equal_memcmp_is_safe_helper<_Elem, _Elem, equal_to<_Elem>> =
+    _Equal_memcmp_is_safe_helper<_Elem, _Elem, equal_to<>>;
 
-template <class _Iter1, class _Iter2, class _Pr>
-false_type _Equal_memcmp_is_safe(const _Iter1&, const _Iter2&, const _Pr&) {
-    // return equal optimization category for arbitrary iterators
-    return {};
-}
+// return equal optimization category for pointers
+template <class _Elem1, class _Elem2, class _Pr, bool = conjunction_v<is_pointer<_Elem1>, is_pointer<_Elem2>>>
+_INLINE_VAR constexpr bool _Equal_memcmp_is_safe =
+    _Equal_memcmp_is_safe_helper<remove_const_t<remove_pointer_t<_Elem1>>, remove_const_t<remove_pointer_t<_Elem2>>,
+        _Pr>;
 
-template <class _Obj1, class _Obj2, class _Pr>
-typename _Equal_memcmp_is_safe_helper<remove_const_t<_Obj1>, remove_const_t<_Obj2>, _Pr>::type _Equal_memcmp_is_safe(
-    _Obj1* const&, _Obj2* const&, const _Pr&) { // return equal optimization category for pointers
-    return {};
-}
+// return equal optimization category for arbitrary iterators
+template <class _Elem1, class _Elem2, class _Pr>

```

Comment:
In reference to [Stephan's previous comment](https://github.com/microsoft/STL/pull/831#discussion_r426100093), should we rename these template parameters to `_Iter1` and `_Iter2` instead of `_Elem1` and `_Elem2`?

Reply:
Im such a derp... Will change soon

Reply:
Done